### PR TITLE
The new format for Handsontable plus the new tables from latest BWWC spreadsheets

### PR DESCRIPTION
**Pull request by KinanBab**
Manually merged master into branch (in the commit before last).

Disclaimer: there are a bunch of small/useless commits on the 16th. I could not install mongo on the lab machine (no sudo access), so I used a free Amazon Ec2 instance for testing. I used the repo to transfer code quickly between the local machine where I was editing, the ec2 instance where I was testing. Sorry for the horrible commit history.

### CODE REVIEW

File: client/index.html [hunk 1]
```diff
@@ -115,8 +116,19 @@
...
+  <div id="compensation-hot"></div>
+  <h3 id="performance-pay-hot-name"></h3>
+  <div id="performance-pay-hot"></div>
+  <h3 id="service-length-hot-name"></h3>
+  <div id="service-length-hot"></div>
+  <h3 id="totals-hot-name"></h3>
+  <div id="totals-hot"></div>
+  
+  <!--
   <div id="board">
```
Comment (frederickjansen) [inline client/index.html]: Remove commented out block

File: client/index.html [hunk 2]
```diff
@@ -128,55 +140,57 @@
...
+        <p class="question-text">Does your company hold events to celebrate diversity and inclusion?</p>
         <div class="radio">
           <label><input autocomplete="off" type="radio" name="optradio" value="Yes">Yes</label>
         </div>
         <div class="radio">
           <label><input autocomplete="off" type="radio" name="optradio" value="No">No</label>
         </div>
         <div class="radio">
-          <label><input autocomplete="off" type="radio" name="optradio" value="NA">Do not wish to answer</label>
+          <label><input autocomplete="off" type="radio" name="optradio" value="NA" checked='checked'>Do not wish to answer</label>
```
Comment (frederickjansen) [inline client/index.html]: Double quotes

File: client/index.html [hunk 3]
```diff
@@ -197,30 +211,70 @@
...
+        // Custom validator that computes the totals
+        var afterChange = function(table) {
+          var index_map = {
+            'Number Of Employees': 0, 
+            'Total Annual Compensation (Dollars)': 1, 
+            'Total Annual Cash Performance Pay (Dollars)': 1, 
+            'Total Length of Service (Months)': 2
+          }
+            
+          var table_index = index_map[table.__sail_meta.name];
```
Comment (frederickjansen) [inline client/index.html]: Use one underscore for private vars

File: client/script/client.js [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    var radios = $(questions[q]).find('input[type=radio]');
+    for(var r = 0; r < radios.length; r++)
+      if(radios[r].checked) { thisQuestionIsValid = true; break; }
+    
+    if(!thisQuestionIsValid) { questionsValid = false; break; }
+  }
+  if (!questionsValid) return callback(false, ADD_QUESTIONS_ERR);
+  
+  // Validate tables (callback chaining)
+  var _ = function validate_callback(i) {
```
Comment (frederickjansen) [inline client/script/client.js]: Use IIFE to call function

File: client/script/client.js [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+ * All inputs are valid. Construct JSON objects and send them to the server.
+ */
+function construct_and_send(target_url, tables, la) {
+  // Start loading animation
+  la.start();
+  
+  // Begin constructing the data  
+  var data_submission = { 'questions': {} };
+  
+  var session = $('#sess').val();
```
Comment (frederickjansen) [inline client/script/client.js]: Trim session and email

File: client/script/client.js [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+/**
+ * All inputs are valid. Construct JSON objects and send them to the server.
+ */
+function construct_and_send(target_url, tables, la) {
+  // Start loading animation
+  la.start();
+  
+  // Begin constructing the data  
+  var data_submission = { 'questions': {} };
```
Comment (frederickjansen) [inline client/script/client.js]: No quotes for JS objects

File: client/script/client.js [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+  
+  var session = $('#sess').val();
+  var email = $('#emailf').val();
+  
+  // Add questions data, each question has three parts:
+  //  'YES', 'NO', and 'NA' and each one has value 0 or 1
+  var questions = $('#questions form');
+  for(var q = 0; q < questions.length; q++) {
+    var question_data = {};
+    var question_text = $($(questions[q]).find('.question-text')[0]).text();
```
Comment (frederickjansen) [inline client/script/client.js]: Possible to write this cleaner?

File: client/script/client.js [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+    data_submission['questions'][question_text] = question_data;
+  }
+  
+  // Handle table data, tables are represented as 2D associative arrays
+  // with the first index being the row key, and the second being the column key
+  var tables_data = construct_data_tables(tables);
+  for(var i = 0; i < tables_data.length; i++)
+    data_submission[tables_data[i].name] = tables_data[i].data;
+    
+  console.log(data_submission);
```
Comment (frederickjansen) [inline client/script/client.js]: Remove log statements

File: client/script/client.js [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+  
+  // Secret share / mask the data.
+  var shares = secretShareValues(data_submission);
+  var data = shares['service'];
```
Comment (frederickjansen) [inline client/script/client.js]: Consistent naming of shares (not sure what way to go)

File: client/script/client.js [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+  pkey_request.then(function(public_key) {
+      mask = encryptWithKey(mask, public_key);
+      var submission = {
+        data: data,
+        mask: mask,
+        user: email,
+        session: session
+      };
+      
+      console.log(submission);
```
Comment (frederickjansen) [inline client/script/client.js]: Remove log

File: client/script/client.js [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+      
+      return $.ajax({
+        type: "POST", url: target_url,
```
Comment (frederickjansen) [inline client/script/client.js]: Hardcode url

File: client/script/client.js [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+        data: JSON.stringify(submission), contentType: 'application/json'
+      });
+  }).then(function(response) {
+    var submitTime = new Date();
+    submitEntries.push({time: submitTime, submitted: true});
+    
+    console.log(response);
```
Comment (frederickjansen) [inline client/script/client.js]: Remove log

File: client/script/client.js [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+    
+    alertify.alert("<img src='style/accept.png' alt='Success'>Success!", "Submitted data.");
+    convertToHTML(submitEntries);
+    
+    // Stop loading animation
+    la.stop();
+    return response;
```
Comment (frederickjansen) [inline client/script/client.js]: Don't need this?

File: client/script/client.js [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+  }).catch(function (err) {
+    var submitTime = new Date();
+    submitEntries.push({time: submitTime, submitted: false});
+    
+    console.log(err);
```
Comment (frederickjansen) [inline client/script/client.js]: Remove log

File: shared/mpc.js [hunk 15]
```diff
@@ -103,7 +113,7 @@ function countInvalidShares (data, db) {
   }
   
   // Access fields in JSON object or in DB object.
-  var fields = db ? function(x){return x.fields;} : function(x){return x;};
+  var fields = db ? function(x){ if(x.fields !== undefined) return x.fields; else return x;} : function(x){return x;};
```
Comment (frederickjansen) [inline shared/mpc.js]: Add comment

File: shared/mpc.js [hunk 16]
```diff
@@ -180,37 +206,78 @@ function aggregateShares (data, db) {
...
-  var agg = {};
-  for (var key in fields(data[0])) {
-    agg[key] = _uint32(0);
-  }
-  for (var i = 0; i < data.length; i++) {
-    for (let key in agg) {
-      var value = convert(fields(data[i])[key]);
-      agg[key] = _addShares(agg[key], value);
+  
+  // TODO FIX WEIRD BEHAVIOR
```
Comment (frederickjansen) [inline shared/mpc.js]: Remove TODO

File: shared/mpc.js [hunk 17]
```diff
@@ -180,37 +206,78 @@ function aggregateShares (data, db) {
...
+  // initialize the aggregate object according to the template
+  var agg = function initialize(obj) {
```
Comment (frederickjansen) [inline shared/mpc.js]: Merge with invalidCount function

File: shared/sail_hot.js [hunk 18]
```diff
@@ -0,0 +1,423 @@
+var HOT_DEFAULT_WIDTH = 1400;
+
+var validators_map = {
```
Comment (frederickjansen) [inline shared/sail_hot.js]: More comments please

File: shared/sail_hot.js [hunk 19]
```diff
@@ -0,0 +1,423 @@
...
+      meta_table.cells[r][c].max = null;
+      meta_table.cells[r][c].min = null;
+    }
+  }
+  
+  table_hot_obj.render();
+}
+
+/**
+ * Change the read only attribute of the entire table.
```
Comment (frederickjansen) [inline shared/sail_hot.js]: Change comment

File: shared/sail_hot.js [hunk 20]
```diff
@@ -0,0 +1,423 @@
...
+    // Workaround for handsontable undo issue for readOnly tables
+    beforeChange: function (changes, source) { return !(this.readOnly); }
+  };
+  
+  // Create the Handsontable
+  var handsOnTable = new Handsontable(element, hotSettings);
+  handsOnTable.__sail_meta = table;
+  
+  // Put name in the title element (if it exists)
+  $('#' + table.element + "-name").html(table.name);
```
Comment (frederickjansen) [inline shared/sail_hot.js]: Remove jQuery dependency

**Review by frederickjansen — CHANGES_REQUESTED** [review]

**Comment by KinanBab** [comment]
@frederickjansen Addressed all comments.

**Pull request merged**